packages/webapp/src/components/Task/AddSoilAmendmentProducts/QuantityApplicationRate/useQuantityApplicationRate.ts

```diff
@@ -0,0 +1,186 @@
+/*
+ *  Copyright 2024 LiteFarm.org
+ *  This file is part of LiteFarm.
+ *
+ *  LiteFarm is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  LiteFarm is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
+ */
+
+import { useEffect, useState } from 'react';
+import { UseFormWatch, UseFormSetValue, UseFormGetValues } from 'react-hook-form';
+import { TASK_PRODUCT_FIELD_NAMES, TaskProductFormFields, UnitOption } from '../types';
+import { getUnitOptionMap } from '../../../../util/convert-units/getUnitOptionMap';
+import { convertFn, getDefaultUnit, location_area } from '../../../../util/convert-units/unit';
+import { roundToTwoDecimal } from '../../../../util';
+import { convert } from '../../../../util/convert-units/convert';
+
+interface UseQuantityApplicationRate {
+  total_area: number;
+  total_area_unit: 'm2' | 'ha' | 'ft2' | 'ac'; // as defined in location_area
+  system: 'metric' | 'imperial';
+  watch: UseFormWatch<TaskProductFormFields>;
+  setValue: UseFormSetValue<TaskProductFormFields>;
+  getValues: UseFormGetValues<TaskProductFormFields>;
+}
+
+export const useQuantityApplicationRate = ({
+  total_area,
+  total_area_unit,
+  system,
+  watch,
+  setValue,
+  getValues,
+}: UseQuantityApplicationRate) => {
+  const application_area = watch(TASK_PRODUCT_FIELD_NAMES.APPLICATION_AREA);
+  const application_area_unit = watch(TASK_PRODUCT_FIELD_NAMES.APPLICATION_AREA_UNIT);
+  const weight_unit = watch(TASK_PRODUCT_FIELD_NAMES.WEIGHT_UNIT);
+  const volume_unit = watch(TASK_PRODUCT_FIELD_NAMES.VOLUME_UNIT);
+  const application_rate_weight_unit = watch(TASK_PRODUCT_FIELD_NAMES.APPLICATION_RATE_WEIGHT_UNIT);
+  const application_rate_volume_unit = watch(TASK_PRODUCT_FIELD_NAMES.APPLICATION_RATE_VOLUME_UNIT);
+
+  const [isWeight, setIsWeight] = useState(() => {
+    const volumeValue = getValues(TASK_PRODUCT_FIELD_NAMES.VOLUME);
+    return isNaN(Number(volumeValue));
+  });
+
+  const toggleMeasure = () => {
+    setIsWeight((prev) => !prev);
```

Comment:
Sorry I forgot to mention that I did go with the third option for toggle behaviour regarding what we looked at in tech daily yesterday (i.e. letting both WEIGHT and VOLUME be saved within their respective components). It just saves so much code relative to converting to maintain the input.

However, if this is the UX we go with, I think this means `isWeight` will need to be added to (and later removed from) the form fields so that the selected value can be identified and the other one cleared at the point of form submission. I'm not sure whether that is preferable to nulling all fields on toggle, so right now neither is implemented, but please let me know which you think is better and of course if there is another option I'm forgetting.

Reply:
I think it's good! (I prefer whatever option results in less code for now since Loïc is away)
I would like to move up `isWeight` without adding to the form fields if possible.